src/basic-syntax/compound-types.md

```diff
@@ -24,3 +24,39 @@ fn main() {
     println!("2nd index: {}", t.1);
 }
 ```
+
+<details>
+    
+Key points:
+    
+Arrays:
+    
+*Arrays have elements of the same type, T, and length, N, which is fixed. 
+
+*We can use literals to assign values to arrays.
+
+*Demonstrate out of bounds errors by setting n to different values 
+
+```
+a[n+15] = 11 // index out of bounds error since len is 10
+```
+
+*Efficient way to check n is in bounds:
```

Comment:
The `n` should be in backticks here — but a bigger question is why you say this is _efficient_? That is, to check if `n` is within bounds, you do `n < a.len()`.

Oh, are you perhaps thinking of the case where you have code like this:

```rust
fn foo(n: usize, a: &[i32]) -> i32 {
    let mut sum = 0;
    for i in 0..20 {
        sum += a[n + i];
    }
    sum
}
```

Here, the Rust compiler will emit 20 bounds checks — it emits one for each of `n + 0`, `n + 1`, ..., `n + 20`. It does this because it wants to panic with an accurate message if, say, `n + 7` is out of bounds.

In this case, you can add

```rust
assert!(n + 20 < a.len());
```

to the top of the function and you will see that subsequent bounds checks go away. It's a nice trick to know and I would be happy to explain it in the speaker notes.

Hmm... actually, I just tried reproducing this and I don't see the promised change in the assembly output. I made a post about it on the [Internals Forum](https://internals.rust-lang.org/t/indexing-into-a-slice-with-a-known-length/18126.), so hopefully someone there will be able to help clear this up.

Reply:
@SkiFire13 [recommended on the Internal Forum](https://internals.rust-lang.org/t/indexing-into-a-slice-with-a-known-length/18126/2?u=mgeisler) using a sub-slice to get great optimization. We could add a note about this "trick" later, it's be very interesting for students who wonder how the performance compares between Rust and C/C++.